File: backend/routes/customer.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { authMiddleware } from '../auth.js';
import { run, get, all } from '../database.js';

const router = express.Router();

// GET /customer/dashboard - Get dashboard summary
router.get('/dashboard', authMiddleware, async (req, res) => {
  try {
    const customerId = req.customerId;

    // Total Requests
    const totalRequestsResult = await get(
      'SELECT COUNT(*) as count FROM sales_requests WHERE customer_id = ?',
      [customerId]
    );
    const totalRequests = totalRequestsResult?.count || 0;

    // Pending Requests (excluding those with Quotation Received or later)
    const pendingRequestsResult = await get(
      `SELECT COUNT(*) as count FROM sales_requests 
       WHERE customer_id = ? AND status IN ('Submitted', 'Under Review')`,
      [customerId]
    );
    const pendingRequests = pendingRequestsResult?.count || 0;

    // Quotations Received
    const quotationsReceivedResult = await get(
      `SELECT COUNT(DISTINCT q.id) as count FROM quotations q
       JOIN sales_requests sr ON q.request_id = sr.id
       WHERE sr.customer_id = ? AND q.quotation_status = 'Awaiting Customer Response'`,
      [customerId]
    );
    const quotationsReceived = quotationsReceivedResult?.count || 0;

    // Quotations Awaiting Customer Action
    const quotationsAwaitingResult = await get(
      `SELECT COUNT(DISTINCT q.id) as count FROM quotations q
       JOIN sales_requests sr ON q.request_id = sr.id
       WHERE sr.customer_id = ? AND q.quotation_status IN ('Awaiting Customer Response', 'Counter Offer')`,
      [customerId]
    );
    const quotationsAwaitingAction = quotationsAwaitingResult?.count || 0;

    // Discount Requests
    const discountRequestsResult = await get(
      `SELECT COUNT(*) as count FROM discount_requests
       WHERE customer_id = ? AND status IN ('Pending Review', 'Requires Manager Approval')`,
      [customerId]
    );
    const discountRequests = discountRequestsResult?.count || 0;

    // Accepted Quotations
    const acceptedQuotationsResult = await get(
      `SELECT COUNT(*) as count FROM quotation_acceptances
       WHERE customer_id = ?`,
      [customerId]
    );
    const acceptedQuotations = acceptedQuotationsResult?.count || 0;

    res.json({
      totalRequests,
      pendingRequests,
      quotationsReceived,
      quotationsAwaitingAction,
      discountRequests,
      acceptedQuotations
    });
  } catch (error) {
    console.error('Dashboard error:', error);
    res.status(500).json({ error: 'Failed to fetch dashboard data' });
  }
});

// GET /customer/requests - Get all requests for the customer
router.get('/requests', authMiddleware, async (req, res) => {
  try {
    const customerId = req.customerId;

    const requests = await all(
      `SELECT id, request_title, product_requirement, quantity, status, created_at
       FROM sales_requests
       WHERE customer_id = ?
       ORDER BY created_at DESC`,
      [customerId]
    );

    // Format dates
    const formattedRequests = requests.map(req => ({
      ...req,
      created_at: new Date(req.created_at).toLocaleDateString('en-IN')
    }));

    res.json(formattedRequests);
  } catch (error) {
    console.error('Fetch requests error:', error);
    res.status(500).json({ error: 'Failed to fetch requests' });
  }
});

// POST /customer/requests - Create a new sales request
router.post('/requests', authMiddleware, async (req, res) => {
  try {
    const customerId = req.customerId;
    const { requestTitle, productRequirement, quantity, specifications, additionalNotes, expectedDeliveryDate } = req.body;

    // Validation
    if (!requestTitle || !productRequirement || !quantity) {
      return res.status(400).json({ error: 'Request title, product requirement, and quantity are required' });
    }

    if (quantity <= 0 || !Number.isInteger(quantity)) {
      return res.status(400).json({ error: 'Quantity must be a positive integer' });
    }

    const requestId = `REQ-${String(Date.now()).slice(-6)}`;

    await run(
      `INSERT INTO sales_requests 
       (id, customer_id, request_title, product_requirement, quantity, specifications, additional_notes, expected_delivery_date, status)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'Submitted')`,
      [requestId, customerId, requestTitle, productRequirement, quantity, specifications || '', additionalNotes || '', expectedDeliveryDate || '']
    );

    res.status(201).json({
      message: 'Request created successfully',
      request: {
        id: requestId,
        requestTitle,
        productRequirement,
        quantity,
        status: 'Submitted',
        createdAt: new Date().toLocaleDateString('en-IN')
      }
    });
  } catch (error) {
    console.error('Create request error:', error);
    res.status(500).json({ error: 'Failed to create request' });
  }
});

// GET /customer/requests/:requestId - Get request details
router.get('/requests/:requestId', authMiddleware, async (req, res) => {
  try {
    const customerId = req.customerId;
    const { requestId } = req.params;

    const request = await get(
      `SELECT * FROM sales_requests WHERE id = ? AND customer_id = ?`,
      [requestId, customerId]
    );

    if (!request) {
      return res.status(404).json({ error: 'Request not found' });
    }

    // Get associated quotation if exists
    const quotation = await get(
      `SELECT * FROM quotations WHERE request_id = ?`,
      [requestId]
    );

    let quotationDetails = null;
    if (quotation) {
      const lineItems = await all(
        `SELECT * FROM quotation_line_items WHERE quotation_id = ?`,
        [quotation.id]
      );

      // Calculate totals
      const subtotal = lineItems.reduce((sum, item) => sum + item.subtotal, 0);
      const totalDiscount = lineItems.reduce((sum, item) => sum + item.discount_amount, 0);
      const totalTax = lineItems.reduce((sum, item) => sum + item.tax_amount, 0);
      const total = lineItems.reduce((sum, item) => sum + item.total_amount, 0);

      quotationDetails = {
        id: quotation.id,
        status: quotation.quotation_status,
        lineItems,
        subtotal: subtotal.toFixed(2),
        totalDiscount: totalDiscount.toFixed(2),
        totalTax: totalTax.toFixed(2),
        total: total.toFixed(2),
        validUntil: quotation.valid_until,
        notes: quotation.notes,
        createdAt: new Date(quotation.created_at).toLocaleDateString('en-IN')
      };
    }

    res.json({
      ...request,
      createdAt: new Date(request.created_at).toLocaleDateString('en-IN'),
      quotation: quotationDetails
    });
  } catch (error) {
    console.error('Fetch request details error:', error);
    res.status(500).json({ error: 'Failed to fetch request details' });
  }
});

// GET /customer/profile - Get customer profile
router.get('/profile', authMiddleware, async (req, res) => {
  try {
    const customer = await get(
      'SELECT id, full_name, company_name, email, phone_number, created_at FROM customers WHERE id = ?',
      [req.customerId]
    );

    if (!customer) {
      return res.status(404).json({ error: 'Customer not found' });
    }

    res.json({
      id: customer.id,
      fullName: customer.full_name,
      companyName: customer.company_name,
      email: customer.email,
      phoneNumber: customer.phone_number,
      createdAt: new Date(customer.created_at).toLocaleDateString('en-IN')
    });
  } catch (error) {
    console.error('Fetch profile error:', error);
    res.status(500).json({ error: 'Failed to fetch profile' });
  }
});

export default router;

File: pybackend/app/routers/customer.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import func, and_
from decimal import Decimal
from datetime import datetime

from app.core.database import get_db
from app.dependencies.auth import get_current_customer
from app.models.customer import Customer
from app.models.sales_request import SalesRequest
from app.models.quotation import Quotation
from app.models.quotation_line import QuotationLineItem
from app.models.discount_request import DiscountRequest
from app.schemas.customer import (
    CreateSalesRequestRequest, SalesRequestResponse, CustomerDashboardResponse,
    CustomerProfileResponse, QuotationResponse, QuotationLineItemResponse,
    RequestDiscountRequest, QuotationSummaryResponse
)
from app.schemas.auth import MessageResponse

router = APIRouter(prefix="/customer", tags=["customer"])

# ==================== DASHBOARD ====================

@router.get("/dashboard", response_model=CustomerDashboardResponse)
def get_customer_dashboard(
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get customer dashboard with summary metrics"""
    customer_id = current["customer_id"]
    
    total_requests = db.query(SalesRequest).filter(SalesRequest.customer_id == customer_id).count()
    pending_requests = db.query(SalesRequest).filter(
        SalesRequest.customer_id == customer_id,
        SalesRequest.status.in_(["Submitted", "Under Review"])
    ).count()
    
    quotations = db.query(Quotation).filter(Quotation.customer_id == customer_id).all()
    quotations_received = len(quotations)
    quotations_awaiting = len([q for q in quotations if q.quotation_status == "Awaiting Customer Response" and not q.accepted])
    accepted_quotations = len([q for q in quotations if q.accepted])
    
    discount_requests = db.query(DiscountRequest).filter(
        DiscountRequest.customer_id == customer_id,
        DiscountRequest.status.in_(["Pending Review", "Requires Manager Approval"])
    ).count()
    
    return CustomerDashboardResponse(
        totalRequests=total_requests,
        pendingRequests=pending_requests,
        quotationsReceived=quotations_received,
        quotationsAwaitingAction=quotations_awaiting,
        discountRequests=discount_requests,
        acceptedQuotations=accepted_quotations
    )

# ==================== PROFILE ====================

@router.get("/profile", response_model=CustomerProfileResponse)
def get_customer_profile(
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get customer profile information"""
    customer = current["customer"]
    
    return CustomerProfileResponse(
        id=customer.id,
        fullName=customer.full_name,
        companyName=customer.company_name,
        email=customer.email,
        phoneNumber=customer.phone_number,
        createdAt=customer.created_at.isoformat() if customer.created_at else None,
        created_at=customer.created_at.isoformat() if customer.created_at else None
    )

# ==================== SALES REQUESTS ====================

@router.post("/requests", response_model=dict, status_code=201)
def create_sales_request(
    request_data: CreateSalesRequestRequest,
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Create a new sales request"""
    customer_id = current["customer_id"]
    
    try:
        sales_request = SalesRequest(
            customer_id=customer_id,
            request_title=request_data.requestTitle,
            product_requirement=request_data.productRequirement,
            quantity=request_data.quantity,
            specifications=request_data.specifications,
            additional_notes=request_data.additionalNotes,
            expected_delivery_date=request_data.expectedDeliveryDate,
            status="Submitted"
        )
        
        db.add(sales_request)
        db.commit()
        db.refresh(sales_request)
        
        return {
            "request": {
                "id": sales_request.id,
                "request_title": sales_request.request_title,
                "product_requirement": sales_request.product_requirement,
                "quantity": sales_request.quantity,
                "specifications": sales_request.specifications,
                "additional_notes": sales_request.additional_notes,
                "expected_delivery_date": sales_request.expected_delivery_date,
                "status": sales_request.status,
                "created_at": sales_request.created_at.isoformat() if sales_request.created_at else None
            }
        }
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to create request")

@router.get("/requests", response_model=list)
def get_customer_requests(
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get all sales requests for customer"""
    customer_id = current["customer_id"]
    
    requests = db.query(SalesRequest).filter(SalesRequest.customer_id == customer_id).all()
    
    result = []
    for req in requests:
        result.append({
            "id": req.id,
            "request_title": req.request_title,
            "product_requirement": req.product_requirement,
            "quantity": req.quantity,
            "specifications": req.specifications,
            "additional_notes": req.additional_notes,
            "expected_delivery_date": req.expected_delivery_date,
            "status": req.status,
            "created_at": req.created_at.isoformat() if req.created_at else None,
            "createdAt": req.created_at.isoformat() if req.created_at else None
        })
    
    return result

@router.get("/requests/{request_id}", response_model=dict)
def get_customer_request_detail(
    request_id: str,
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get details of a specific sales request"""
    customer_id = current["customer_id"]
    
    sales_request = db.query(SalesRequest).filter(
        SalesRequest.id == request_id,
        SalesRequest.customer_id == customer_id
    ).first()
    
    if not sales_request:
        raise HTTPException(status_code=404, detail="Request not found")
    
    quotation = db.query(Quotation).filter(Quotation.request_id == request_id).first()
    
    response = {
        "id": sales_request.id,
        "request_title": sales_request.request_title,
        "product_requirement": sales_request.product_requirement,
        "quantity": sales_request.quantity,
        "specifications": sales_request.specifications,
        "additional_notes": sales_request.additional_notes,
        "expected_delivery_date": sales_request.expected_delivery_date,
        "status": sales_request.status,
        "created_at": sales_request.created_at.isoformat() if sales_request.created_at else None,
        "createdAt": sales_request.created_at.isoformat() if sales_request.created_at else None
    }
    
    if quotation:
        response["quotation"] = {
            "id": quotation.id,
            "status": quotation.quotation_status,
            "total": str(quotation.total_amount),
            "validUntil": quotation.valid_until,
            "createdAt": quotation.created_at.isoformat() if quotation.created_at else None,
            "accepted": quotation.accepted,
            "acceptedAt": quotation.accepted_at.isoformat() if quotation.accepted_at else None
        }
    
    return response

# ==================== QUOTATIONS ====================

@router.get("/quotations", response_model=list)
def get_customer_quotations(
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get all quotations for customer"""
    customer_id = current["customer_id"]
    
    quotations = db.query(Quotation).filter(Quotation.customer_id == customer_id).all()
    
    result = []
    for q in quotations:
        # Get associated request for title
        req = db.query(SalesRequest).filter(SalesRequest.id == q.request_id).first()
        
        result.append({
            "id": q.id,
            "requestTitle": req.request_title if req else "",
            "quantity": req.quantity if req else 0,
            "status": q.quotation_status,
            "validUntil": q.valid_until,
            "createdAt": q.created_at.isoformat() if q.created_at else None,
            "requestId": q.request_id
        })
    
    return result

@router.get("/quotations/{quotation_id}", response_model=dict)
def get_customer_quotation_detail(
    quotation_id: str,
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Get details of a specific quotation"""
    customer_id = current["customer_id"]
    
    quotation = db.query(Quotation).filter(
        Quotation.id == quotation_id,
        Quotation.customer_id == customer_id
    ).first()
    
    if not quotation:
        raise HTTPException(status_code=404, detail="Quotation not found")
    
    # Get line items
    line_items = db.query(QuotationLineItem).filter(QuotationLineItem.quotation_id == quotation_id).all()
    
    line_items_list = []
    for item in line_items:
        line_items_list.append({
            "product_name": item.product_name,
            "quantity": item.quantity,
            "unit_price": str(item.unit_price),
            "subtotal": str(item.subtotal),
            "discount_percent": float(item.discount_percent),
            "discount_amount": str(item.discount_amount),
            "tax_amount": str(item.tax_amount),
            "total_amount": str(item.total_amount)
        })
    
    # Get discount requests
    discount_requests_list = []
    discount_requests = db.query(DiscountRequest).filter(DiscountRequest.quotation_id == quotation_id).all()
    for dr in discount_requests:
        discount_requests_list.append({
            "id": dr.id,
            "requestedDiscount": float(dr.requested_discount_percent),
            "currentDiscount": float(dr.current_discount_percent) if dr.current_discount_percent else 0,
            "reason": dr.reason,
            "status": dr.status,
            "createdAt": dr.created_at.isoformat() if dr.created_at else None,
            "salespersonResponse": dr.salesperson_response
        })
    
    return {
        "id": quotation.id,
        "requestId": quotation.request_id,
        "status": quotation.quotation_status,
        "total": str(quotation.total_amount),
        "subtotal": str(quotation.subtotal),
        "totalDiscount": str(quotation.total_discount),
        "totalTax": str(quotation.total_tax),
        "validUntil": quotation.valid_until,
        "notes": quotation.notes,
        "accepted": quotation.accepted,
        "acceptedAt": quotation.accepted_at.isoformat() if quotation.accepted_at else None,
        "lineItems": line_items_list,
        "discountRequests": discount_requests_list
    }

@router.post("/quotations/{quotation_id}/discount-request", response_model=dict, status_code=200)
def request_quotation_discount(
    quotation_id: str,
    discount_request: RequestDiscountRequest,
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Request a discount on a quotation"""
    customer_id = current["customer_id"]
    
    quotation = db.query(Quotation).filter(
        Quotation.id == quotation_id,
        Quotation.customer_id == customer_id
    ).first()
    
    if not quotation:
        raise HTTPException(status_code=404, detail="Quotation not found")
    
    try:
        disc_req = DiscountRequest(
            quotation_id=quotation_id,
            customer_id=customer_id,
            salesperson_id=quotation.salesperson_id,
            requested_discount_percent=Decimal(str(discount_request.requestedDiscountPercent)),
            current_discount_percent=quotation.discount_percent,
            reason=discount_request.reason,
            customer_message=discount_request.customerMessage,
            status="Pending Review",
            requires_manager_approval=discount_request.requestedDiscountPercent > float(quotation.salesperson.max_discount_percent)
        )
        
        db.add(disc_req)
        db.commit()
        db.refresh(disc_req)
        
        return {
            "message": "Discount request submitted",
            "discountRequest": {
                "id": disc_req.id,
                "requestedDiscount": float(disc_req.requested_discount_percent),
                "currentDiscount": float(disc_req.current_discount_percent) if disc_req.current_discount_percent else 0,
                "reason": disc_req.reason,
                "status": disc_req.status,
                "createdAt": disc_req.created_at.isoformat() if disc_req.created_at else None
            }
        }
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to submit discount request")

@router.post("/quotations/{quotation_id}/accept", response_model=dict)
def accept_quotation(
    quotation_id: str,
    current: dict = Depends(get_current_customer),
    db: Session = Depends(get_db)
):
    """Accept a quotation"""
    customer_id = current["customer_id"]
    
    quotation = db.query(Quotation).filter(
        Quotation.id == quotation_id,
        Quotation.customer_id == customer_id
    ).first()
    
    if not quotation:
        raise HTTPException(status_code=404, detail="Quotation not found")
    
    try:
        quotation.accepted = True
        quotation.accepted_at = datetime.utcnow()
        quotation.quotation_status = "Accepted"
        
        # Update associated request status
        sales_request = db.query(SalesRequest).filter(SalesRequest.id == quotation.request_id).first()
        if sales_request:
            sales_request.status = "Accepted"
        
        db.commit()
        
        return {
            "message": "Quotation accepted successfully",
            "quotation": {
                "id": quotation.id,
                "status": quotation.quotation_status,
                "accepted": quotation.accepted,
                "acceptedAt": quotation.accepted_at.isoformat() if quotation.accepted_at else None
            }
        }
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail="Failed to accept quotation")
